File: alg/math/gcd.c
/*************************************************************************
    > File Name: gcd.c
    > Author: hao
    > Mail: onerhao@gmail.com
    > Created Time: Thu 22 Nov 2012 02:20:14 PM CST
 ************************************************************************/
int gcd(int x, int y)
{
    int t;
    while (y > 0)
    {
        if (x > y)
        {
            t = x % y;
            x = y;
            y = t;
        }
        else
        {
            y %= x;
        }
    }
    return x;
}

int gcdn(int *a, int n)
{ // return the greatest common divisor of n numbers in array a
    int t, i;
    if (n < 1)
        return 0;
    t = gcd(a[0], a[1]);
    for (i = 2; i < n; i++)
    {
        t = gcd(t, a[i]);
    }
    return t;
}

File: alg/math/gcd.py
# -*- coding:utf-8 -*-
'''
Greatest Common Divisor,Euclid Algorithm
'''


class gcd(object):

    """gcd algorithm"""

    def __init__(self):
        """@todo: to be defined1. """
        pass

    def euclid(self, a, b):
        if b == 0:
            return a
        else:
            return self.euclid(b, a % b)

    def extended_euclid(self, a, b):
        """ extended gcd algorithm,returning the x,y,
        for which gcd(a,b) = ax + by
        """
        if b == 0:
            return (a, 1, 0)
        else:
            d1, x1, y1 = self.extended_euclid(b, a % b)
            d, x, y = d1, y1, x1 - (a / b) * y1
            return (d, x, y)

if __name__ == "__main__":
    print gcd().euclid(30, 21)
    print gcd().extended_euclid(99, 78)
